### ISSUE TITLE
MySQL Tests fail in docker environment

### ISSUE THREAD

**Issue by nniehoff** [nautobot/nautobot#1516]
### Environment
* Python version: 3.7
* Nautobot version: 1.2.9 (develop)

### Steps to Reproduce
1.  Clone the Nautobot repo, checkout develop
2.  invoke destroy
3.  switch to the MySQL database backend.  (#1514 can help with this)
4. remove an local invoke.yml file and `INVOKE_*` environment variables
5. invoke unittest

### Expected Behavior

Tests succeed

### Observed Behavior

Tests fail

<details>

```no-highlight
[2022-03-17 16:07:35,779: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,799: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,828: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,867: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,901: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,935: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:35,973: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
[2022-03-17 16:07:36,074: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/d162655d-0f5d-4ce5-9291-bdecf5f7ec8f/available-ips/
.[2022-03-17 16:07:36,936: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/1c4b39bb-256f-4a55-bf7c-d141345f0f72/available-prefixes/
[2022-03-17 16:07:36,959: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/1c4b39bb-256f-4a55-bf7c-d141345f0f72/available-prefixes/
[2022-03-17 16:07:36,979: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/1c4b39bb-256f-4a55-bf7c-d141345f0f72/available-prefixes/
[2022-03-17 16:07:37,035: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/1c4b39bb-256f-4a55-bf7c-d141345f0f72/available-prefixes/
[2022-03-17 16:07:37,085: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/1c4b39bb-256f-4a55-bf7c-d141345f0f72/available-prefixes/
.
======================================================================
FAIL: test_query_devices_filter (nautobot.core.tests.test_graphql.GraphQLQueryTest) [Checking q: "first"] (filter='q: "first"', nbr_expected_results=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/core/tests/test_graphql.py", line 1041, in test_query_devices_filter
    self.assertEqual(len(result.data["devices"]), nbr_expected_results)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.DeviceTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 1686, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.InventoryItemTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 3139, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.RackTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 577, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_search (nautobot.extras.tests.test_filters.ExportTemplateTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/extras/tests/test_filters.py", line 310, in test_search
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 3)
AssertionError: 0 != 3

======================================================================
FAIL: test_device_mac_address_icontains (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 612, in test_device_mac_address_icontains
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 2)
AssertionError: 0 != 2

======================================================================
FAIL: test_device_mac_address_icontains_negation (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 616, in test_device_mac_address_icontains_negation
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 1)
AssertionError: 3 != 1

======================================================================
FAIL: test_device_mac_address_startswith (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 596, in test_device_mac_address_startswith
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_device_mac_address_startswith_negation (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 600, in test_device_mac_address_startswith_negation
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 2)
AssertionError: 3 != 2

----------------------------------------------------------------------
Ran 4679 tests in 1019.193s

FAILED (failures=9, skipped=74)
```

</details>

**Comment by glennmatthews** [comment]
I haven't looked at every one of the failures in depth but at least the first three in the list all deal with case-insensitive matches. It's possible that our recommended MySQL collation setting of `utf8mb4_bin` is to blame here - we might try switching to `utf8mb4_general_ci` instead and see if that fixes the issue?

**Comment by nniehoff** [comment]
@glennmatthews that seemed to fix the issue with the tests, PR incoming.

**Comment by jathanism** [comment]
Food for thought on `utf8mb4_general_ci` vs. `utf8mb4_unicode_ci`: https://stackoverflow.com/a/766996

**Comment by jathanism** [comment]
Here are the test failures I experienced running Nautobot 1.2.11:

```
$ invoke unittest
Running docker-compose command "ps --services --filter status=running"
Running docker-compose command "run --entrypoint 'coverage run --module nautobot.core.cli --config=nautobot/core/tests/nautobot_config.py test nautobot --buffer' nautobot"
[+] Running 3/0
 ⠿ Container nautobot-selenium-1  Running                                                     0.0s
 ⠿ Container nautobot-redis-1     Running                                                     0.0s
 ⠿ Container nautobot-db-1        Running                                                     0.0s
Wrapping model clean methods for custom validators failed because the ContentType table was not available or populated. This is normal during the execution of the migration command for the first time.
Creating test database for alias 'default'...

System check identified some issues:

WARNINGS:
<Settings "nautobot_config">: (nautobot.core.W005) STORAGE_CONFIG has been set but STORAGE_BACKEND is not defined. STORAGE_CONFIG will be ignored.

System check identified 1 issue (0 silenced).
...................s.........................................s....................s........................................................................................................................................................................................................s.......................................................s....................s.....................s....................s.....................s....................s....................s..............................................s....................s....................s....................s....................s.....................s....................s....................s.........................................s....................s.....................s....................s....................s.....................s....................s....................s..............................................s....................s....................s................................................s............................................................................................F..................................................................F.......................................................................................F...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................s....................s.......................s............................s....s...............s....................s.........................s....s................s................s......................s.....................s...............s....................s.............s..................s....................s....................s....................s....................s....s...............s...................................................................................F.........................................................................................................................................................s...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................s......................................................................s....................s....................s.....................s....................s..........................................s....................................................................................................................................................s..................................................................................................................................................................................................................................................................................................................................................s....................s.......................................................................s...............s....................s......s...............s.....................................................FF.FF..................................................................................................................................s....................s....................s....................s..........................................................................................................................................................................................................................................[2022-04-06 17:19:43,416: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,473: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,517: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,590: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,610: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,738: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,852: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
[2022-04-06 17:19:43,985: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/17feb5a6-a5ef-496c-94da-c6f88595e2ef/available-ips/
.[2022-04-06 17:19:47,650: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/98a7e61a-b73b-42b7-8d92-d2122de45146/available-prefixes/
[2022-04-06 17:19:47,779: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/98a7e61a-b73b-42b7-8d92-d2122de45146/available-prefixes/
[2022-04-06 17:19:47,891: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/98a7e61a-b73b-42b7-8d92-d2122de45146/available-prefixes/
[2022-04-06 17:19:48,017: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/98a7e61a-b73b-42b7-8d92-d2122de45146/available-prefixes/
[2022-04-06 17:19:48,150: ERROR/MainProcess] Internal Server Error: /api/ipam/prefixes/98a7e61a-b73b-42b7-8d92-d2122de45146/available-prefixes/
.
======================================================================
FAIL: test_query_devices_filter (nautobot.core.tests.test_graphql.GraphQLQueryTest) [Checking q: "first"] (filter='q: "first"', nbr_expected_results=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/core/tests/test_graphql.py", line 1306, in test_query_devices_filter
    self.assertEqual(len(result.data["devices"]), nbr_expected_results)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.DeviceTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 1686, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.InventoryItemTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 3139, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_serial (nautobot.dcim.tests.test_filters.RackTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/dcim/tests/test_filters.py", line 577, in test_serial
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_search (nautobot.extras.tests.test_filters.ExportTemplateTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/extras/tests/test_filters.py", line 310, in test_search
    self.assertEqual(self.filterset(params, self.queryset).qs.count(), 3)
AssertionError: 0 != 3

======================================================================
FAIL: test_device_mac_address_icontains (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 612, in test_device_mac_address_icontains
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 2)
AssertionError: 0 != 2

======================================================================
FAIL: test_device_mac_address_icontains_negation (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 616, in test_device_mac_address_icontains_negation
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 1)
AssertionError: 3 != 1

======================================================================
FAIL: test_device_mac_address_startswith (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 596, in test_device_mac_address_startswith
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_device_mac_address_startswith_negation (nautobot.utilities.tests.test_filters.DynamicFilterLookupExpressionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/source/nautobot/utilities/tests/test_filters.py", line 600, in test_device_mac_address_startswith_negation
    self.assertEqual(DeviceFilterSet(params, self.device_queryset).qs.count(), 2)
AssertionError: 3 != 2

----------------------------------------------------------------------
Ran 4700 tests in 3550.046s
```
FAILED (failures=9, skipped=74)
Destroying test database for alias 'default'...
```

**Comment by jathanism** [comment]
Ok, the fix is complicated. 

- Do not create databases using `utf8mb4_bin` collation because this breaks case-sensitive queries. Let the database default to the server default of `utf8mb4_0900_ai_ci` (case-insensitive). Because this the MySQL default, [Django also expects this to be the case](https://docs.djangoproject.com/en/4.0/ref/databases/): 

> By default, with a UTF-8 database, MySQL will use the utf8_general_ci collation. This results in all string equality comparisons being done in a case-insensitive manner. That is, "Fred" and "freD" are considered equal at the database level. If you have a unique constraint on a field, it would be illegal to try to insert both "aa" and "AA" into the same column, since they compare as equal (and, hence, non-unique) with the default collation. If you want case-sensitive comparisons on a particular column or table, change the column or table to use the utf8_bin collation.

- We need `CollateAsChar` to continue to use `utf8mb4_bin` for `Interface.name` and `VMInterface.name` as was already the case prior and we want these queries to be case-sensitive for ONLY these columns
- Update the test config to always explicitly set the client charset to `utf8mb4` (4-byte encoding) as MySQL versions < 8.0.28 default to using `utf8mb3` (3-byte encoding)

**Comment by jathanism** [comment]
Closed by #1528
